human-tracking-and-pose-detection/main.py

```diff
@@ -0,0 +1,217 @@
+import argparse
+import cv2
+import math
+import numpy as np
+import tensorflow as tf
+import tensorflow_hub as hub
+import time
+parser_args = tuple()
+
+
+def parse_arguments():
+    """
+    handle command line arguments
+    """
+    parser = argparse.ArgumentParser(description='Example command-line parser')
+    parser.add_argument(
+        '-v',
+        '--verbose',
+        action='store_true',
+        help='Enable print_verbose_only outputs')
+    parser.add_argument(
+        '-k',
+        '--display_keypoint',
+        action='store_true',
+        help='Enable display of keypoints')
+    parser.add_argument(
+        '-i',
+        '--display_image',
+        action='store_true',
+        help='Enable display of image and keypoints')
+    return parser.parse_args()
+
+
+def print_verbose_only(*args, **kwargs):
+    """
+    print only if verbose==True
+    """
+    if parser_args.verbose:
+        print(*args, **kwargs)
+
+
+def print_fps(func):
+    """
+    function decorator for printing fps
+    """
+    def wrapper(*args, **kwargs):
+        prev_time = time.time()
+        result = func(*args, **kwargs)
+        cur_time = time.time()
+        if cur_time - prev_time > (0.000001):
+            fps = 1.0 / (cur_time - prev_time)
+            print_verbose_only('FPS: {:.2f}'.format(fps))
+        return result
+    return wrapper
+
+
+def movenet(input_image, model):
+    """
+    movenet model:
+    Gets input image and outputs array of keypoints with certainty score
+    downloaded from #https://tfhub.dev/google/movenet/multipose/lightning/1
+    """
+    # SavedModel format expects tensor type of int32.
+    input_image = tf.cast(input_image, dtype=tf.int32)
+    outputs = model(input_image)  # Output is a [1, 6, 56] tensor.
+
+    keypoints = outputs['output_0'].numpy()
+
+    threshold = 0.3
+    count_of_people = np.sum(keypoints[0, :, -1] > threshold)
+    print_verbose_only("count_of_people", count_of_people)
+    return keypoints[:, :, :51].reshape((6, 17, 3))[0]
+
+
+def isTherePerson(points, score_threshold=0.3):
+    """
+    return True/False of whether there is a person
+    """
+    visible_joints = np.sum(points[:, 2] > score_threshold)
+    return visible_joints >= 3
+
+
+def isPersonFacingCamera(points, score_threshold=0.3):
+    """
+    return True/False depending on if the person is facing camera or not
+    """
+    LEFT_EYE = 1
+    NOSE = 0
+    RIGHT_EYE = 2
+    visible_joints_face = np.sum(points[:5, 2] > score_threshold)
+    facing_forward = points[LEFT_EYE][1] > points[NOSE][1] > points[RIGHT_EYE][1]
+    return visible_joints_face >= 3 and facing_forward
+
+
+def getHeadingAngle(
+        points,
+        score_threshold=0.3,
+        fov=90,
+        image_width=1,
+        offset=0,
+        scaling=1):
```

Comment:
What are the units here?

Reply:
there are no units. i think They are constants (that will be) derived by testing on the trailbot, so that the coordinates are accurate with realworld measurements